polaris-react/src/components/DataTable/DataTable.tsx

```diff
@@ -795,18 +801,21 @@ class DataTableInner extends PureComponent<CombinedProps, DataTableState> {
               inStickyHeader,
             });
           }}
-          inFixedNthColumn
+          inFixedNthColumn={Boolean(fixedFirstColumns)}
+          lastFixedFirstColumn={headingIndex === fixedFirstColumns - 1}
           style={{
             left: this.state.columnVisibilityData[headingIndex]?.leftEdge,
-            borderRight:
-              headingIndex === fixedFirstColumns - 1 && fixedCellVisible
-                ? 'var(--p-border-divider)'
-                : undefined,
```

Comment:
It was impossible to get the borders to be in the exact same position in the different contexts this way. I had to introduce `lastFixedFirstColumn` to be able to use `CSS` `::after` pseudo-element which gave more reliable results.